src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <cmath>
+#include <fstream>
+#include <map>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "../../../../runtime/file_utils.h"
+#include "../../../../target/source/codegen_c.h"
+
+namespace tvm {
+namespace runtime {
+
+using namespace tir;
+
+class CodeGenCMSISNN : public tvm::codegen::CodeGenC {
+ public:
+  void Init(bool output_ssa) {
+    decl_stream << "#include <stdio.h>\n";
+    decl_stream << "#include <stdlib.h>\n";
+    decl_stream << "#include <dlpack/dlpack.h>\n";
+    decl_stream << "#include <tvm/runtime/crt/module.h>\n";
+    decl_stream << "#include <arm_nnfunctions.h>\n";
+    CodeGenC::Init(output_ssa);
+  }
+
+  /*!
+   * \brief Emit code that offloads a subgraph to the Cortex-M
+   *
+   * \return string of code that offloads a subgraph to the Cortex-M
+   */
+  void AddFunction(const PrimFunc& prim_func) {
+    PrintExternCPrefix(stream);
+    CodeGenC::AddFunction(prim_func);
+    PrintExternCPostfix(stream);
+  }
+
+ private:
+  void VisitExpr_(const CallNode* op, std::ostream& os) {  // NOLINT(*)
+    if (!op->op.same_as(builtin::call_extern())) {
+      return;
+    }
+    std::string cmsis_func_name = op->args[0].as<StringImmNode>()->value;
+    if (cmsis_func_name == "arm_softmax_s8") {
+      EmitSoftmax(op);
```

Comment:
Can we not use the default behaviour now for `arm_softmax_s8` due to it emitting a standard `call_extern`? I'd still leave the infrastructure in for the struct calls but I don't think we need it specifically for this call?

Reply:
What I mean is that I think we can delete the `VisitExpr` and other overrides but leave the class in place for extension with the struct methods?

Reply:
I would like to leave it as is to maintain consistency when new operators come in.

Reply:
I would suggest we don't duplicate code unnecessarily, softmax can always use the default codegen infrastructure?

Reply:
It generates the plain call without leaving details of params that makes the parameter reading harder.

It generates something like this:
arm_softmax_s8(input, 256, 3, 1099511680, 25, -62, output);

vs.

what is being generated now:
  int32_t num_rows = 256;
  int32_t row_size = 3;
  int32_t mult = 1099511680;
  int32_t shift = 25;
  int32_t diff_min = -62;
  arm_softmax_s8(input,   num_rows, row_size, mult, shift, diff_min, output);


I would prefer the latter. wdyt?

Reply:
Given we have full documentation for the signature of `arm_softmax_s8` and these variables replicated in the TIR generation - I would suggest we have sufficient information on the arguments already.

Reply:
Removed the VisitExpr_ for Softmax.